pkg/clustermanager/kube_proxy.go

```diff
@@ -0,0 +1,505 @@
+		return err
+	}
+
+	// Delete original kube-proxy pods to ensure there is only one copy of kube-proxy running
+	// on each node.
+	if err := deleteAllOriginalKubeProxyPods(ctx, client); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func iptablesLegacyKubeProxyFromCurrentDaemonSet(kcp *controlplanev1.KubeadmControlPlane, kubeProxy *appsv1.DaemonSet) *appsv1.DaemonSet {
+	iptablesLegacyKubeProxy := kubeProxy.DeepCopy()
+
+	// Generate a new DS with the old kube-proxy version with nodeAffinity so it only
+	// gets scheduled in the old (current) nodes.
+	iptablesLegacyKubeProxy.Name = iptablesLegacyKubeProxyDSName
+	iptablesLegacyKubeProxy.ObjectMeta.ResourceVersion = ""
+	iptablesLegacyKubeProxy.ObjectMeta.UID = ""
+	image := kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.ImageRepository + "/kube-proxy" +
+		":" + kcp.Spec.Version
+	iptablesLegacyKubeProxy.Spec.Template.Spec.Containers[0].Image = image
+	iptablesLegacyKubeProxy.Spec.Template.Spec.Affinity = &corev1.Affinity{
+		NodeAffinity: &corev1.NodeAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
+				NodeSelectorTerms: []corev1.NodeSelectorTerm{
+					{
+						MatchExpressions: []corev1.NodeSelectorRequirement{
+							{
+								Key:      iptablesLegacyLabel,
+								Operator: corev1.NodeSelectorOpExists,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return iptablesLegacyKubeProxy
+}
+
+func createIPTablesLegacyKubeProxy(ctx context.Context, client client.Client, kcp *controlplanev1.KubeadmControlPlane, originalKubeProxy *appsv1.DaemonSet) error {
+	iptablesLegacyKubeProxy := iptablesLegacyKubeProxyFromCurrentDaemonSet(kcp, originalKubeProxy)
+	if err := client.Create(ctx, iptablesLegacyKubeProxy); err != nil && !apierrors.IsAlreadyExists(err) {
+		return errors.Wrap(err, "creating secondary kube-proxy DS with iptables-legacy for old nodes")
+	}
+
+	return nil
+}
+
+func deleteIPTablesLegacyKubeProxy(ctx context.Context, client client.Client) error {
+	iptablesLegacyKubeProxy := &appsv1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      iptablesLegacyKubeProxyDSName,
+			Namespace: kubeProxyDSNamespace,
+		},
+	}
+
+	if err := client.Delete(ctx, iptablesLegacyKubeProxy); err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrap(err, "deleting secondary kube-proxy DS with iptables-legacy")
+	}
+
+	return nil
+}
+
+func updateKubeProxyVersion(ctx context.Context, client client.Client, kubeProxy *appsv1.DaemonSet, image string) error {
+	kubeProxy.Spec.Template.Spec.Containers[0].Image = image
+	if err := client.Update(ctx, kubeProxy); err != nil {
+		return errors.Wrap(err, "updating main kube-proxy version before upgrade")
+	}
+
+	return nil
+}
+
+func (u KubeProxyUpgrader) ensureUpdateKubeProxyVersion(ctx context.Context, log logr.Logger, client client.Client, spec *cluster.Spec) error {
+	newKubeProxyImage := spec.VersionsBundle.KubeDistro.KubeProxy.URI
+	kubeProxy, err := getKubeProxy(ctx, client)
```

Comment:
why needing the first getKubeProxy, updateKubeProxyVersion separately from the retrier?

Reply:
We don't need it. It avoids having the retrier always fail on the first iteration.
That's not a big deal so I changed the logging a bit to make it more explicit and simplified the code. Let me know.